### PR TITLE
Integrate Accelerate, Deepspeed

### PR DESCRIPTION
**Pull request by Adamliu1**
Partial Fix for #115, #115 will be deprecated, tasks inside #115 will be split into multiple PRs.

This PR contains code for issues #105 #113

Here are the following todos and changes made:
- [x] Integrate Accelerator (commented out bf16 currently, can be used as an argument in the future, so far, using full precision)
- [x] Fix model saving after distributed training
- [x] Fix logging, including default logging and wandb. But use Accelerator's wrapper
  - [x] wandb
  - [x] logging
- [x] Currently commented out mink
- [x] Add documentation about installing and use accelerate, deepspeed

### CODE REVIEW

File: llm_unlearn_ucl/unlearn_harm.py [hunk 1]
```diff
@@ -324,19 +353,20 @@ def main(args) -> None:
...
-    )
-    if os.path.isfile(normal_sample_path):
-        data_sample_artifacts.add_file(
-            normal_sample_path,
-            name=f"normal_{args.samples_count}_samples.json",
-        )
-    if os.path.isfile(bad_sample_path):
-        data_sample_artifacts.add_file(
-            bad_sample_path, name=f"bad_{args.samples_count}_samples.json"
+    if bool(args.wandb_log) and accelerator.is_local_main_process:
```
Comment (Davidyz) [inline llm_unlearn_ucl/unlearn_harm.py:356]: The uploading of dataset samples should go [here](https://github.com/Adamliu1/SNLP_GCW/blob/380f2e755a7893f77a80c8adab6d4f461d2a2c1e/llm_unlearn_ucl/data_utils.py#L66), as suggested in #106 .

**Review by Davidyz — CHANGES_REQUESTED** [review]

File: llm_unlearn_ucl/logger_utils.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    if os.path.isfile(normal_sample_path):
+        data_sample_artifacts.add_file(
+            normal_sample_path, name=f"normal_{args.samples_count}_samples.json"
+        )
+    if os.path.isfile(bad_sample_path):
+        data_sample_artifacts.add_file(
+            bad_sample_path, name=f"bad_{args.samples_count}_samples.json"
+        )
+    wandb_tracker = accelerator.get_tracker("wandb", unwrap=True)
+    wandb_tracker.log_artifact(data_sample_artifacts)
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/logger_utils.py:32]: I am unsure if we still need this in the codebase, given we never used those logs (they were supposed for reproducibility) and it adds additional processing time per batch. 

Maybe let's remove?

File: llm_unlearn_ucl/unlearn_harm.py [hunk 3]
```diff
@@ -54,64 +57,64 @@ def set_seed(seed_num: int) -> None:
...
-            probability = probabilities[s_idx, i - 1, token_id].item()
-            all_prob.append(probability)
-        # Get top-K % probs and compute their mean (it was log_softmax, so top k% is actually mink% prob)
-        k_length = int(len(all_prob) * K)
-        topk_prob = np.sort(all_prob)[:k_length]
-        pred_mink.append(-np.mean(topk_prob).item())
-
-    # All mean MIN-K% prob in: pred_mink. For now, return all
-    return pred_mink
+# def compute_mink_prob(
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:60]: Please remove, we no longer need it. In the worst case we will look back in git history :)

🎶For the first time in **history**
It's gonna start raining men (Start raining men)! 🎶

File: llm_unlearn_ucl/unlearn_harm.py [hunk 4]
```diff
@@ -129,40 +132,43 @@ def run_training_batch(
...
-    )
-    # Calculate min-k% prob score on normal_batch using the model under unlearning
-    mink_probs_after_step_normal = compute_mink_prob(
-        model=model,
-        batch=normal_batch,
-        K=args.mink_prob_k,
-        device=device,
-        compute_for_answer_only=False,
-    )
+    # # Calculate min-k% prob score on bad_batch using the unmodified pre-trained model
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:138]: Same here, please remove.

File: llm_unlearn_ucl/unlearn_harm.py [hunk 5]
```diff
@@ -190,20 +197,20 @@ def run_training_batch(
...
                 "bad_loss": -bad_loss,
                 "normal_loss": normal_loss,
                 "final_loss": loss,
-                "ratio (bad) mink unlearning/reference": np.mean(mink_probs_after_step)
-                / np.mean(mink_probs_base),
-                "ratio (normal) mink unlearning/reference": np.mean(
-                    mink_probs_after_step_normal
-                )
-                / np.mean(mink_probs_base_normal),
+                # "ratio (bad) mink unlearning/reference": np.mean(mink_probs_after_step)
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:208]: please remove.

File: llm_unlearn_ucl/unlearn_harm.py [hunk 6]
```diff
@@ -212,11 +219,12 @@ def run_training_batch(
         f"samples seen: {samples_count}, "
         f"bad_loss: {-bad_loss:.2f}, "
         f"current_div_loss: {normal_loss:.2f}, "
-        f"ratio (bad) mink unlearning/reference: {np.mean(mink_probs_after_step)/np.mean(mink_probs_base):.3f}, "
-        f"ratio (normal) mink unlearning/reference: {np.mean(mink_probs_after_step_normal)/np.mean(mink_probs_base_normal):.3f}"
+        # f"ratio (bad) mink unlearning/reference: {np.mean(mink_probs_after_step)/np.mean(mink_probs_base):.3f}, "
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:222]: please remove.

File: llm_unlearn_ucl/unlearn_harm.py [hunk 7]
```diff
@@ -229,25 +237,47 @@ def main(args) -> None:
     assert (
         args.samples_count // args.sequential
     ) % args.batch_size == 0, "samples in each 'sequence' (--samples_count / --sequential) should be a multiple of batch_size."
-    accelerator = Accelerator(
-        mixed_precision="bf16"
-    )  # accelerator precision can be specified if required.
+
+    if args.wandb_log:
+        accelerator = Accelerator(log_with="wandb", mixed_precision="bf16")
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:242]: For visibility: @Willmish's code removes bf16 from the python code and defines it in the deepspeed config.

File: llm_unlearn_ucl/unlearn_harm.py [hunk 8]
```diff
@@ -229,25 +237,47 @@ def main(args) -> None:
...
+        accelerator.init_trackers(
+            project_name=args.wandb_project_name,
+            config=vars(args),
+            init_kwargs={"wandb": {"name": args.wandb_run_name}},
+        )
+    else:
+        accelerator = Accelerator(mixed_precision="bf16")
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:249]: For visibility: @Willmish's code removes bf16 from the python code and defines it in the deepspeed config.

File: llm_unlearn_ucl/unlearn_harm.py [hunk 9]
```diff
@@ -229,25 +237,47 @@ def main(args) -> None:
...
+
     print(f"Loading model {args.model_name} for training...")
     if args.use_quantized:
         # Uncomment for quantized
         model = AutoModelForCausalLM.from_pretrained(
             args.model_name,
             cache_dir=args.cache_dir,
             load_in_8bit=True,
-            torch_dtype=torch.float32,
+            torch_dtype=torch.bfloat16,
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:273]: For visibility: @Willmish's code removes bf16 from the python code and defines it in the deepspeed config.

File: llm_unlearn_ucl/unlearn_harm.py [hunk 10]
```diff
@@ -229,25 +237,47 @@ def main(args) -> None:
...
         )
     else:
         model = AutoModelForCausalLM.from_pretrained(
             args.model_name,
             cache_dir=args.cache_dir,
             torch_dtype=torch.bfloat16,
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:279]: For visibility: @Willmish's code removes bf16 from the python code and defines it in the deepspeed config.

File: llm_unlearn_ucl/unlearn_harm.py [hunk 11]
```diff
@@ -372,27 +405,32 @@ def main(args) -> None:
...
 
-    print(f"Loading model {args.model_name} for reference ('fully learned')...")
     if args.use_quantized:
         # Uncomment for quantized
         pretrained_model = AutoModelForCausalLM.from_pretrained(
             args.model_name,
             cache_dir=args.cache_dir,
             load_in_8bit=True,
-            torch_dtype=torch.float32,
+            torch_dtype=torch.bfloat16,
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:417]: For visibility: @Willmish's code removes bf16 from the python code and defines it in the deepspeed config.

File: llm_unlearn_ucl/unlearn_harm.py [hunk 12]
```diff
@@ -372,27 +405,32 @@ def main(args) -> None:
...
+            trust_remote_code=True,
         )
     else:
         pretrained_model = AutoModelForCausalLM.from_pretrained(
             args.model_name,
             cache_dir=args.cache_dir,
             torch_dtype=torch.bfloat16,
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:424]: For visibility: @Willmish's code removes bf16 from the python code and defines it in the deepspeed config.

File: llm_unlearn_ucl/unlearn_harm.py [hunk 13]
```diff
@@ -412,12 +449,16 @@ def main(args) -> None:
...
             epoch_num = 0
             accu_bad_loss = 0
+            bad_loss = 0
             while epoch_num < args.num_epochs:
                 accu_bad_loss = 0
                 for normal_batch, bad_batch in zip(
                     train_normal_loader, train_bad_loader
                 ):
                     samples_count += len(bad_batch["input_ids"])
+                    # TODO: verify if this implementation is correct
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:459]: @Adamliu1 ???

File: llm_unlearn_ucl/unlearn_harm.py [hunk 14]
```diff
@@ -523,7 +572,9 @@ def main(args) -> None:
 
             epoch_num += 1
 
+            # NOTE: here need to verify logic
```
Comment (TheRootOf3) [inline llm_unlearn_ucl/unlearn_harm.py:575]: @Adamliu1, is this sorted?

**Review by TheRootOf3 — CHANGES_REQUESTED** [review]
good job, comments left 🫡

**Review by TheRootOf3 — APPROVED** [review]
lgtm

**Pull request merged**